common/src/main/java/com/microsoft/identity/common/internal/activebrokerdiscovery/BrokerDiscoveryClientFactory.kt

```diff
@@ -0,0 +1,79 @@
+//  Copyright (c) Microsoft Corporation.
+//  All rights reserved.
+//
+//  This code is licensed under the MIT License.
+//
+//  Permission is hereby granted, free of charge, to any person obtaining a copy
+//  of this software and associated documentation files(the "Software"), to deal
+//  in the Software without restriction, including without limitation the rights
+//  to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+//  copies of the Software, and to permit persons to whom the Software is
+//  furnished to do so, subject to the following conditions :
+//
+//  The above copyright notice and this permission notice shall be included in
+//  all copies or substantial portions of the Software.
+//
+//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+package com.microsoft.identity.common.internal.activebrokerdiscovery
+
+import android.content.Context
+import com.microsoft.identity.common.java.interfaces.IPlatformComponents
+import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+
+/**
+ * A class for initializing a new [IBrokerDiscoveryClient] object.
+ **/
+class BrokerDiscoveryClientFactory {
+    companion object {
+        @Volatile
+        private var IS_NEW_DISCOVERY_ENABLED = false
+
+        @Volatile
+        private var instance: IBrokerDiscoveryClient? = null
+
+        private val lock = Mutex()
+
+        /**
+         * If set to true, the new Broker discovery mechanism will be enabled.
+         * This is currently turned off by default - until we're ready to ship the feature.
+         **/
+        @JvmStatic
+        fun setNewBrokerDiscoveryEnabled(isEnabled: Boolean){
+            if (isEnabled != IS_NEW_DISCOVERY_ENABLED) {
```

Comment:
I'm not sure I understand what we're trying to do here. Why can't we just set the flag?

Reply:
if the flag is changed, then i would need to wipe the existing singleton (which would belong to a different type)